chain/chain/src/flat_storage_resharder.rs

```diff
@@ -396,8 +405,74 @@ impl FlatStorageResharder {
         // Terminate the resharding event.
         *self.resharding_event.lock().unwrap() = None;
     }
+
+    /// Returns an iterator over a shard's flat storage at the given block hash. This
+    /// iterator contains both flat storage values and deltas.
+    fn flat_storage_iterator<'a>(
+        &self,
+        flat_store: &'a FlatStoreAdapter,
+        shard_uid: &ShardUId,
+        block_hash: &CryptoHash,
+    ) -> Result<Box<FlatStorageAndDeltaIter<'a>>, Error> {
+        let mut iter: Box<FlatStorageAndDeltaIter<'a>> = Box::new(
+            flat_store
+                .iter(*shard_uid)
+                // Get the flat storage iter and wrap the value in Optional::Some to
+                // match the delta iterator so that they can be chained.
+                .map_ok(|(key, value)| (key, Some(value)))
+                // Wrap the iterator's item into an Entry.
+                .map(|entry| FlatStorageAndDeltaIterItem::Entry(entry)),
+        );
+
+        // Get all the blocks from flat head to the wanted block hash.
+        let flat_storage = self
+            .runtime
+            .get_flat_storage_manager()
+            .get_flat_storage_for_shard(*shard_uid)
+            .expect("the flat storage undergoing resharding must exist!");
+        // Must reverse the result because we want ascending block heights.
+        let mut blocks_to_head = flat_storage.get_blocks_to_head(block_hash).map_err(|err| {
+            StorageError::StorageInconsistentState(format!(
+                "failed to find path of blocks to flat storage head ({err})"
+            ))
+        })?;
+        blocks_to_head.reverse();
+        debug!(target = "resharding", "flat storage blocks to head len = {}", blocks_to_head.len());
+
+        // Get all the delta iterators and wrap the items in Result to match the flat
+        // storage iter so that they can be chained.
+        for block in blocks_to_head {
+            let deltas = flat_store.get_delta(*shard_uid, block).map_err(|err| {
+                StorageError::StorageInconsistentState(format!(
+                    "can't retrieve deltas for flat storage at {block}/{shard_uid:?}({err})"
+                ))
+            })?;
+            let Some(deltas) = deltas else {
+                continue;
+            };
+            // Chain the iterators effectively adding a block worth of deltas.
+            // Before doing so insert a commit point to separate changes to the same key in different transactions.
```

Comment:
Nice, I wouldn't have thought of that. How did you find that it's needed?

Reply:
Tests failed with transaction overwrite error 😉